packages/vm/src/limited.rs

```diff
@@ -0,0 +1,135 @@
+//! A set of tools designed for processing user defined contract data,
+//! which can potientially have abusive size.
+
+use std::collections::{BTreeSet, HashSet};
+use std::iter::FromIterator;
+
+pub trait LimitedDisplay {
+    /// Returns a string representationof the object, which is shorter than or equal to `max_length`.
+    /// Implementations must panic if `max_length` is not reasonably large.
+    fn to_string_limited(&self, max_length: usize) -> String;
+}
+
+impl<E: Ord + AsRef<str>> LimitedDisplay for BTreeSet<E> {
+    fn to_string_limited(&self, max_length: usize) -> String {
+        let mut out = String::with_capacity(max_length * 130 / 100);
+        let opening = "{";
+        let closing = "}";
+
+        let mut first = true;
+        out.push_str(opening);
+        let mut lengths_stack = Vec::<usize>::new();
+        for element in self.iter() {
+            lengths_stack.push(out.len());
+
+            if first {
+                out.push('"');
+                first = false;
+            } else {
+                out.push_str(", \"");
+            }
+            out.push_str(element.as_ref());
+            out.push('"');
```

Comment:
Maybe I'm missing something, but this looks needlessly complicated to me.

Why not something like:
```suggestion
            if first {
                out.push('"');
                first = false;
            } else if out.len() + 3 + element.len() + 1 + closing.len() <= max_len {
                out.push_str(", \"");
                out.push_str(element.as_ref());
                out.push('"');
            }
```

That way you don't need to store the lenghts, backtrack later, etc.

Reply:
Yeah, exactly. The problem is I don't even know the length of the rest string in advance.